### PR TITLE
Add Data Analysis Scripts

### LINKED ISSUE #3: Write script for extracting stats of output json file
Write some scripts (python is also good for this) that gets the `output.json` and extracts the stats about the simulation result.

### PR DESCRIPTION
**Pull request by BTreadwell**

### CODE REVIEW

File: simulator/dataAnalysis/dataAnalysis.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        if event['kind'] == 'add-block':
+            block_id = event['content']['block-id']
+            node_id = event['content']['node-id']
+
+            # if this block has been seen by a node before
+            if block_id in block_propagation.keys():
+                # add the new node in the list of receivers
+                block_propagation[block_id].add(node_id)
+                # the first time the block is seen by all nodes, mark the propagation end timestamp
+                if len(block_propagation[block_id]) == num_nodes and len(block_propagation_times[block_id]) < 2:
```
Comment (hosseins) [inline simulator/dataAnalysis/dataAnalysis.py:41]: is `len(block_propagation_times[block_id])==1` better? and why do you need to check that?

File: simulator/dataAnalysis/dataAnalysis.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                # create the entry, we use a set because we don't want to store duplicates
+                block_propagation[block_id] = {node_id}
+                # log the propagation start time, we use list because we want to maintain order
+                block_propagation_times[block_id] = [event['content']['timestamp']]
+
+    # assertion: block_propagation_times is now a hashtable with keys consisting of each block id
+    # and a value which is the propagation start timestamp and a propagation end timestamp
+    propagation_times = [block_propagation_times[x][1] - block_propagation_times[x][0] for x in
+                             block_propagation_times.keys()]
+    propagation_times.sort()
```
Comment (hosseins) [inline simulator/dataAnalysis/dataAnalysis.py:53]: You can directly get min and max over an array with a cost of O(n). Sorting is O(log(n)) at least.

File: simulator/dataAnalysis/dataAnalysis.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    min_propagation_time = propagation_times[0]
+    max_propagation_time = propagation_times[-1]
+    # odd length
+    if len(propagation_times) & 1:
```
Comment (hosseins) [inline simulator/dataAnalysis/dataAnalysis.py:57]: you can equally use statistics library for median:
https://www.geeksforgeeks.org/find-median-of-list-in-python/

**Review by josedeveloper2026 — COMMENTED** [review]
Thanks! Overall looks good. I added some comments.

File: simulator/dataAnalysis/dataAnalysis.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    block_propagation = {}
+    for event in json_list:
+        if event['kind'] == 'add-block':
+            block_id = event['content']['block-id']
+            node_id = event['content']['node-id']
+
+            # if this block has been seen by a node before
+            if block_id in block_propagation.keys():
+                # add the new node in the list of receivers
+                block_propagation[block_id].add(node_id)
```
[the remark this answers is not in the document]
Comment (BTreadwell) [reply simulator/dataAnalysis/dataAnalysis.py:39]: this is already using a set

File: simulator/dataAnalysis/dataAnalysis.py (hunk 1 above)
Comment (BTreadwell) [reply simulator/dataAnalysis/dataAnalysis.py:41]: Yes; switched to ==1. I believe a block can be sent to a node that has already seen it, so we want to log the first end time. I'll double check that that is the case.

Comment (hosseins) [reply simulator/dataAnalysis/dataAnalysis.py:41]: That assumption makes sense, so its good to have ==1

File: simulator/dataAnalysis/dataAnalysis.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hosseins) [reply simulator/dataAnalysis/dataAnalysis.py:39]: I see it now. Thanks!

**Review by josedeveloper2026 — APPROVED** [review]
LGTM! This is great thanks!
Please merge and squash this branch to the main.

**Pull request merged**